File: C#/Arrays/AssociativeArray/AssociativeArray.cs
/*******************************************************
 *  @file ArrayList.ccs
 *  @author Stephen Hall
 *  @date 11/17/17.
 *  @copyright 2017 Stephen Hall. All rights reserved.
 *  @details ArrayList implementation in C#
 ********************************************************/

namespace DataStructures.Arrays.AssociativeArray
{
    /// <summary>
    /// Associative Array class declaration
    /// </summary>
    /// <typeparam name="TKey"></typeparam>
    /// <typeparam name="TValue"></typeparam>
    public class AssociativeArray<TKey, TValue>
    {

        /// <summary>
        /// Node class declaration
        /// </summary>
        public class Node
        {
            public TKey Key;
            public TValue Value;
            public Node Next;
            public int Hash;

            /// <summary>
            /// Node class Constructor
            /// </summary>
            /// <param name="key">key of the node</param>
            /// <param name="value">value for node to hold</param>
            /// <param name="hash">hash index for index</param>
            public Node(TKey key, TValue value, int hash)
            {
                Key = key;
                Value = value;
                Hash = hash;
            }

        }

        private Node[] _table;
        private int _size;

        /// <summary>
        /// AssociativeArray class Constructor
        /// </summary>
        public AssociativeArray()
        {
            _table = new Node[10];
            _size = 0;
        }

        /// <summary>
        ///  AssociativeArray class Constructo
        /// </summary>
        /// <param name="size">Size to initialize array to</param>
        public AssociativeArray(int size)
        {
            _table = new Node[size];
            _size = 0;
        }

        /// <summary>
        /// Adds or updates Key, Value pair into the Array
        /// </summary>
        /// <param name="key">Key to associate vale with</param>
        /// <param name="value">value to store</param>
        /// <returns>Node added or updated</returns>
        public Node Set(TKey key, TValue value)
        {
            // Find the hash of the key and bucket it belongs to 
            int hash = key.GetHashCode();
            int bucket = GetBucket(hash);
            Node entry;
            if (IsEmpty())
            {
                entry = new Node(key, value, hash);
                _table[bucket] = entry;
                _size++;
            }
            else
            {
                entry = _table[bucket];
                while (entry.Next != null)
                {
                    if (entry.GetHashCode() == hash && entry.Key.Equals(key))
                    {
                        entry.Value = value;
                        return entry;
                    }
                    entry = entry.Next;
                }

                Node node = new Node(key, value, hash);
                entry.Next = node;
                _size++;
                entry = node;
            }
            return entry;
        }

        /// <summary>
        /// Gets the value of the given key
        /// </summary>
        /// <param name="key">Key to get value of</param>
        /// <returns>value of the key</returns>
        public TValue Get(TKey key)
        {
            int hash = key.GetHashCode();
            int bucket = GetBucket(hash);

            Node entry = _table[bucket];
            while (entry != null)
            {
                /* If hash and key matches, return the value */
                if ((entry.Hash == hash) && entry.Key.Equals(key))
                {
                    return entry.Value;
                }
                entry = entry.Next;
            }
            return default(TValue);
        }

        /// <summary>
        /// Gets the size of the array
        /// </summary>
        /// <returns>number of elements in the array</returns>
        public int Size() => _size;

        /// <summary>
        /// Checks if the array is empty of not
        /// </summary>
        /// <returns>true|false</returns>
        public bool IsEmpty() => _size == 0;

        /// <summary>
        /// Gets the bucket container for the internal array
        /// </summary>
        /// <param name="hash">hash to find bucket of</param>
        /// <returns>bucket index of the array</returns>
        private int GetBucket(int hash) => (hash % _table.Length);
    }
}

File: C++/Arrays/AssociativeArray/AssociativeArray.cpp
/*******************************************************
 *  @file AssociativeArray.cpp
 *  @author Stephen Hall
 *  @date 11/27/17.
 *  @copyright 2017 Stephen Hall. All rights reserved.
 *  @details AssociativeArray implementation in C++
 ********************************************************/

#include <iostream>
#include "AssociativeArray.h"
using namespace Arrays::AssociativeArray;
using namespace std;

/**
 * Associative Array Node constructor
 */
AssociativeArray::Node::Node()
{
	key = NULL;
	value = NULL;
	hash = 0;
	next = NULL;
}

/**
 * Associative Array Node constructor
 * @tparam Key: Generic Type
 * @tparam Value: Generic Type
 * @param key: key of the node
 * @param value: value of the key
 * @param hash: bucket index of the node
 */
template<typename Key, typename Value>
AssociativeArray::Node::Node(Key key, Value value, int hashCode)
{
	this->key = key;
	this->hash = hashCode;
	this->value = value;
	this->next = NULL;
}

/**
 * Associative Array class constructor
 * @tparam Key: Generic type
 * @param hashFunction: Function pointer to user defined hash function for type Key
 */
template<typename Key>
AssociativeArray::AssociativeArray(unsigned int (*hashFunction)(Key))
{
	table = new Node*[size = 10];
	this->hash = hashFunction;
}

/**
 * Associative Array class constructor
 * @tparam Key: Generic type
 * @param size: Size to initialize the array to
 * @param hashFunction: function pointer to user defined hash function for type Key
 */
template<typename Key>
AssociativeArray::AssociativeArray(unsigned int size, unsigned int (*hashFunction)(Key))
{
	table = new Node*[this->size = size];
	this->hash = hashFunction;
}

/**
 * Associative Array class destructor
 */
AssociativeArray::~AssociativeArray()
{
	destroy();
}

/**
 * Overloaded operator to act as an array
 * @tparam Key: Generic type
 * @tparam Value: Generic Type
 * @param key: Key to find
 * @return Value: Value of the key
 */
template<typename Key, typename Value>
Value& AssociativeArray::operator[](Key key)
{
	unsigned int hash = this->hash(key) % size;
	Node* tmp = table[hash];
	while(tmp != NULL)
	{
		if(tmp->key == key)
			return tmp->value;
	}
	return NULL;
}

/**
 * Adds or updates Key, Value pair into the Array
 * @tparam Key: Generic Type
 * @tparam Value: Generic Type
 * @param key: Key to add into the array
 * @param value: Value of the key
 * @return Node*: New node added into the array
 */
template<typename Key, typename Value>
Node* AssociativeArray::set(Key key, Value value)
{
	int hash = hashCode(key);
	int bucket = getBucket(hash);
	Node* entry;
	if(isEmpty()){
		entry = new Node(key, value, hash);
		table[bucket] = entry;
		count++;
	}
	else {
		entry = table[bucket];
		while (entry->next != NULL) {
			/* Key exists, update value */
			if (hashCode(entry->key) == hash && equalTo(entry->key, key)) {
				entry->value = value;
				return entry;
			}
			entry = entry->next;
		}
		/* Add new key value pair into the array */
		Node* node = new Node(key, value, hash);
		entry->next = node;
		count++;
		entry = node;
	}
	return entry;
}

/**
 * Keys the value of the given key
 * @tparam Key: Generic Type
 * @tparam Value: Generic Type
 * @param key: Key to find
 * @return Value: value of the key
 */
template<typename Key, typename Value>
Value AssociativeArray::get(Key key)
{
	int hash = hashCode(key);
	int bucket = getBucket(hash);

	Node* entry = table[bucket];
	while (entry != NULL) {
		/* If hash and key matches, return the value */
		if ((entry->hash == hash) && equalTo(entry->key, key)) {
			return entry->value;
		}
		entry = entry->next;
	}
	return NULL;
}

/**
 * returns the number of elements in the array
 * @return int: number of elements in the array
 */
int AssociativeArray::length()
{
	return count;
}

/**
 * determins if the array is empty
 * @return bool: true|false
 */
bool AssociativeArray::isEmpty()
{
	return size == 0;
}

/**
 * Gets the bucket index of the given hash
 * @param hash: Hash to index
 * @return int: bucket index of the hash
 */
int AssociativeArray::getBucket(int hash)
{

	return hash % size;
}

/**
 * Determines if a is equal to b
 * @tparam Key: Generic type
 * @param a: key to test
 * @param b: second key to test
 * @return bool: true|false
 */
template<typename Key>
bool AssociativeArray::equalTo(Key a, Key b)
{
	return a == b;
}

/**
 * Frees the memory dynamically allocated
 */
void AssociativeArray::destroy()
{
	Node* tmp;
	Node* tmp2;
	for(int i = 0; i < size; i++)
	{
		tmp = table[i];
		tmp2 = tmp;
		while(tmp2 != NULL)
		{
			tmp2 = tmp2->next;
			delete tmp;
			tmp = tmp2;
		}
	}
	delete [] table;
	table = NULL;
}

/**
 * Calls the function pointer for hash code give at time of creation
 * @tparam Key: Generic Type
 * @param key: Key to hash
 * @return unsigned int: Hash of the key
 */
template<typename Key>
unsigned int AssociativeArray::hashCode(Key key)
{
	return this->hash(key);
}
